### ISSUE TITLE
SPIKE!  How is the URL path  parameter ‘application_id’ sanitized?  

### ISSUE THREAD

**Issue by gormless87** [bcgov/nr-forests-access-management#950]
**Describe the task**
Investigate current URL parameter sanitization process and verify secure coding practices are being followed

**Acceptance Criteria**
- [ ] the 5 questions listed under Additional context are unambiguoulsy answered

**Additional context**
Questions from Chris Prince:

Forgive my limited experience with python but I was updating the risks list and thinking about the input sanitization and secure coding items, I have questions –

1.	How is the URL path  parameter ‘application_id’ sanitized?  
2.	Can a string format injection or ASCII escape code sequence be written to the LOGGER?
	In the Model application_id is a BigInteger with maxvalue the application_id parameter is an int which I think is unbound. 
3.	Should the application_id parameter by the same python type in the model?
4.	What happens if an invalid application_id parameter is passed to the database query?
5.	will an exception/error text message from the database layer be sent to the caller, or will the exception/error be caught and a error message used that does not disclose any information about the implementation of the database be used?

**Comment by MCatherine1994** [comment]
1. How is the URL path parameter ‘application_id’ sanitized?

We have two routers have ‘application_id’ parameter. We check the type of the application_id to be an integer.

2. Can a string format injection or ASCII escape code sequence be written to the LOGGER?

I'm not so sure about this. Wondering if @ianliuwk1019 has any idea?

3. In the Model application_id is a BigInteger with maxvalue, in the schema application_id parameter is an int which I think is unbound. Should the application_id parameter by the same python type in the model?

Wondering if we need to use BigInteger in the database for the application_id? Maybe we should keep the model and schema consistent.

From [postgres](https://www.postgresql.org/docs/9.1/datatype-numeric.html):
The type integer is the common choice, as it offers the best balance between range, storage size, and performance. The smallint type is generally only used if disk space is at a premium. The bigint type should only be used if the range of the integer type is insufficient, because the latter is definitely faster.
The INT/INTEGER stores values between -2,147,483,648 to +2,147,483,647. The BIGINT data type stores the values between -9,223,372,036,854,775,808 to +9,223,372,036,854,775,807

4. What happens if an invalid application_id parameter is passed to the database query?

If I pass a string value as the application, I got an invalid type error. 
![image](https://github.com/bcgov/nr-forests-access-management/assets/23131735/7e95f897-c234-4014-a5a2-b213ad05f203)

5. will an exception/error text message from the database layer be sent to the caller, or will the exception/error be caught and a error message used that does not disclose any information about the implementation of the database be used?

I feel the it won't reach to the database layer cause it's block by the type check at the router level?

**Comment by ianliuwk1019** [comment]
For (3.): 
I think initially the model was generated from flyway and for some reason, maybe for the table sequence, probably it is standard (not sure, from DA point of view) to use `bigint`. I feel we don't necessary need to change this to be consistent but we just limit schema object(python) `application_id` to smaller (`int`) than the database limit (`bitint`).
However, surprisingly (I didn't know this chanage), when I checked again, Python2 has `int` limit, but Python3 dropped the `int` limit. So now the `int` limit for Python depends on the platform's `sys.maxsize `. We might need to check what that limit is.
Ref: https://note.nkmk.me/en/python-int-max-value/#:~:text=In%20Python%203%2C%20the%20int,which%20has%20no%20max%20limit.
Ref: https://docs.python.org/3/whatsnew/3.0.html#integers (at PEP238).

**Comment by MCatherine1994** [comment]
1. How is the URL path parameter ‘application_id’ sanitized?

    We have two routers have the ‘application_id’ parameter, `get_fam_application_roles` and `get_fam_application_user_role_assignment`. We check the type of the application_id to be an integer. Later on once we set a limit of the application id, will also check the given application_id value is greater than 0 and less than the limit. 

2. Can a string format injection or ASCII escape code sequence be written to the LOGGER?

    We only accept number format. I didn't find a way that we can give a string value for the application_id in the frontend, because the two router methods which have application_id are called automatically by our application, there is no place for a user to give an input. If try to give a string value and call the api directly, no log will be generated, because the api call is blocked by the router level. FastAPI checks the type error. 

3. In the Model application_id is a BigInteger with maxvalue, in the schema application_id parameter is an int which I think is unbound. Should the application_id parameter by the same python type in the model?

    We feel it's fine to keep the current implementation, and we have another ticket to set limit the of the application_id #988. 

4. What happens if an invalid application_id parameter is passed to the database query?

    If pass a string value as the application_id, will get a type error at the router level, that blocks the api call, won't even run the database query. 

5. will an exception/error text message from the database layer be sent to the caller, or will the exception/error be caught and a error message used that does not disclose any information about the implementation of the database be used?
 
   For two router methods have application_id, we handle the following error cases: 
       - when application_id is invalid -> FastAPI returns type error
       - when application_id not exists -> we capture it and return the error message "Application ID {application_id} not found"
       - if the user makes the api call has no access to this application -> we capture it and return the error message "Operation requires role {required_role}"
